### PR TITLE
enable single quote for rust (correctly)

### PR DESCRIPTION
**Pull request by numToStr**
Ref: https://github.com/windwp/nvim-autopairs/pull/143#issuecomment-930783297

### CODE REVIEW

**Comment by numToStr** [comment]
I am not sure if this is the correct way to add a rule for a specific filetype. Please feel free to correct me.

**Comment by windwp** [comment]
@numToStr yes you can do that maybe add %w too

@felix-clark can you confirm . i need to learn rust first :)

**Comment by numToStr** [comment]
@windwp I suppose I need to add another regex check like this `cond.not_before_regex_check('%w')`?

**Comment by numToStr** [comment]
@windwp I've added the `%w` check :)

**Comment by felix-clark** [comment]
Good point, I totally forgot about single `chars`. I'm not sure this regex will handle *every* case but it's certainly much better. Perhaps allowing an arbitrary length of whitespace after `[&<]` might handle some additional edge cases?

We might also want to disable autopairs after a comma, since I think you can have multiple lifetime annotations at once e.g. `MyStruct<'a, 'b, ...>`. So perhaps `[&<,]` with an optional amount of whitespace following?

**Comment by felix-clark** [comment]
One potential "problem" with disabling after a comma is that there would not be completions in such a case:
```
let a = ['a', 'b', 'c'];
```
However this is probably not common at all since `String`s are a thing. And in my personal opinion, false positives are a lot more annoying than false negatives.

**Comment by numToStr** [comment]
@felix-clark Nice catch. It is possible to have multiple lifetimes and covering them might be hard. As you said, this could be conflicting with `Vec<char>`. Though we can cover a case where the after char is `>`, so if we are at the of the angle bracket and press the `'` quote then it won't create a pair. Thoughts?

**Comment by felix-clark** [comment]
Seems like it would work for the initial typing, assuming the `>` has been inserted properly by autopairs. However it might fail if one was attempting to make the change `MyStruct<'a, T>` -> `MyStruct<'a, 'b, T>` for whatever reason.

I wondered if it might be better to allow an arbitrary length of a certain class of characters between the angle brackets and the cursor, however, it appears that the nvim-autopairs rules API only allows for checking the immediately adjacent character. Perhaps @windwp could confirm this, but within that constraint I think that what you have is pretty close to as good as it can get, and should work as intended for the majority of cases.

**Comment by felix-clark** [comment]
One final comment, perhaps the last ">" check doesn't need to be a regex since it is only a single character. I'd think you could use `not_after_text_check(">")`.

**Comment by numToStr** [comment]
> I wondered if it might be better to allow an arbitrary length of a certain class of characters between the angle brackets and the cursor

I was also wondering about this.

> I'd think you could use `not_after_text_check(">")`

I'll try this.

> And perhaps combine the first two regexes into `[%w&<]`

I already tried and it doesn't seem to be working. Maybe @windwp can come with a regex. I am a noob in lua regex :)

**Comment by numToStr** [comment]
@felix-clark @windwp changes are done. I think it's good to be merged.

File: lua/nvim-autopairs/rules/basic.lua [hunk 1]
```diff
@@ -29,8 +29,10 @@ local function setup(opt)
         Rule('"""', '"""', { 'python', 'elixir' }),
         basic("'", "'")
             :with_pair(cond.not_before_regex_check("%w"))
-            :with_pair(cond.not_filetypes({"rust"}))
-        ,
+            :with_pair(cond.not_filetypes({"rust"})),
+        basic("'", "'", "rust")
+            :with_pair(cond.not_before_regex_check("[%w<&]")),
```
Comment (felix-clark) [inline lua/nvim-autopairs/rules/basic.lua:34]: I'm not very familiar with lua but should there be a comma at the end of this line? It doesn't seem to match the patterns elsewhere in this source file.

Comment (numToStr) [reply lua/nvim-autopairs/rules/basic.lua:34]: You are right. It seems I can't type correctly 😅

**Comment by windwp** [comment]
@numToStr @felix-clark  i think it working now. thank

**Comment by yioneko** [comment]
@felix-clark I think you can use treesitter api to get a smarter pair completion, like this:
```lua
local npairs = require("nvim-autopairs")
local ts_conds = require("nvim-autopairs.ts-conds")

npairs.setup({
    check_ts = true,
    ts_config = {
      rust = {},
    },
  })
npairs.get_rule("'")[2]:with_pair(ts_conds.is_not_ts_node({ "type_arguments", "bounded_type" }))
```
This should handle  `MyStruct<'a, T>` -> `MyStruct<'a, 'b, T>` correctly.

**Comment by yioneko** [comment]
@felix-clark I think you can use treesitter api to get a smarter pair completion, like this:
```lua
local npairs = require("nvim-autopairs")
local ts_conds = require("nvim-autopairs.ts-conds")

npairs.setup({
    check_ts = true,
    ts_config = {
      rust = {},
    },
  })
npairs.get_rule("'")[2]:with_pair(ts_conds.is_not_ts_node({ "type_arguments", "bounded_type" }))
```
This should handle  `MyStruct<'a, T>` -> `MyStruct<'a, 'b, T>` correctly.

**Comment by glyh** [comment]
Is this working? I've tried to type some codes like:
```rust
pub struct Iter<T,
```
when I type `'` next, it autocompletes a pair of `'` for me.

**Pull request merged**